pycbc/inference/gelman_rubin.py

```diff
@@ -0,0 +1,174 @@
+
+    # get end index of blocks
+    ends = numpy.arange(start, end, step)
+    stats = numpy.zeros((nparameters, len(ends)))
+
+    # get start index of blocks
+    starts = numpy.array(len(ends) * [start])
+
+    # loop over end indexes and calculate statistic
+    for i, e in enumerate(ends):
+        tmp = chains[:,:,0:e]
+        stats[:, i] = gelman_rubin(tmp)
+
+    return starts, ends, stats
+
+def gelman_rubin(chains, auto_burn_in=True):
+    """ Calculates the univariate Gelman-Rubin convergence statistic.
+
+    Parameters
+    ----------
+    chains : iterable
+        An iterable of numpy.array instances that contain the samples
+        for each chain. Each chain has shape (nparameters, niterations).
+    auto_burn_in : bool
+        If True, then only use later half of samples provided.
+
+    Returns
+    -------
+    psrf : numpy.array
+        A numpy.array of shape (nparameters) that has the point estimates of
+        the potential scale reduction factor.
+    """
+
+    # remove first have of samples
+    # this will have shape (nchains, nparameters, niterations)
+    if auto_burn_in:
+        _, _, niterations = numpy.array(chains).shape
+        chains = numpy.array([chain[:, niterations / 2 + 1:]
+                              for chain in chains])
+
+    # get number of chains, parameters, and iterations
+    chains = numpy.array(chains)
+    nchains, nparameters, niterations = chains.shape
+
+    # calculate the covariance matrix for each chain
+    # this will have shape (nchains, nparameters, nparameters)
+    chains_covs = numpy.array([numpy.cov(chain) for chain in chains])
+    if nparameters == 1:
+        chains_covs = chains_covs.reshape((nchains, 1, 1))
+
+    # calculate W
+    # this will have shape (nparameters, nparameters)
+    w = numpy.zeros(chains_covs[0].shape)
+    for i, row in enumerate(chains_covs[0]):
+        for j, _ in enumerate(row):
+            w[i, j] = numpy.mean(chains_covs[:, i, j])
+    if nparameters == 1:
+        w = w.reshape((1, 1))
+
+    # calculate B
+    # this will have shape (nparameters, nparameters)
+    means = numpy.zeros((nparameters, nchains))
+    for i, chain in enumerate(chains):
+        means[:, i] = numpy.mean(chain, axis=1).transpose()
+    b = niterations * numpy.cov(means)
+    if nparameters == 1:
+        b = b.reshape((1, 1))
+
+    # get diagonal elements of W and B
+    # these will have shape (nparameters)
+    w_diag = numpy.diag(w)
+    b_diag = numpy.diag(b)
+
+    # get variance for each chain
+    # this will have shape (nparameters, nchains)
+    var = numpy.zeros((nparameters, nchains))
+    for i, chain_cov in enumerate(chains_covs):
+        var[:, i] = numpy.diag(chain_cov)
+
+    # get mean of means
+    # this will have shape (nparameters)
+    mu_hat = numpy.mean(means, axis=1)
+
+    # get variance of variance
+    # this will have shape (nparameters)
+    s = numpy.var(var, axis=1)
+
+    # get factors in variance of V calculation
```

Comment:
It would be helpful in terms of understanding the code if this block (line #139 - #144 ) is moved after lines 146-148, ie. first getting V and then calculating its variance.